src/main/scala/ai/privado/languageEngine/java/tagger/source/SourceTagger.scala

```diff
@@ -0,0 +1,325 @@
+    cpgNodeCache.tagAllFieldAccessAndGetters(builder, typeDeclVal, taggerPart.ruleInfo, typeDeclMember.name, true)
+  }
+}
+
+/** Step 2.1 =>
+  *
+  * Tag identifier of all the typeDeclaration who have a member as memberName in argument Represent Step 2.1
+  */
+class FirstLevelDerivedSourceTagger(
+  cpg: Cpg,
+  cpgNodeCache: CPGNodeCacheForSourceTagger,
+  ruleCache: RuleCache,
+  taggerCache: TaggerCache
+) extends PrivadoParallelCpgPass[DerivedSourceTaggerPart](cpg) {
+  override def generateParts(): Array[DerivedSourceTaggerPart] = cpgNodeCache.derivedSourceTaggerParts
+  override def runOnPart(builder: DiffGraphBuilder, taggerPart: DerivedSourceTaggerPart): Unit = {
+    val typeDeclMember = taggerPart.member
+    val typeDeclVal    = taggerPart.typeDecl.fullName
+    // updating cache
+    taggerCache.addItemToTypeDeclMemberCache(typeDeclVal, taggerPart.ruleInfo.id, typeDeclMember)
+    val typeDeclMemberName = typeDeclMember.name
+    // Have started tagging Parameters as well, as in collection points sometimes there is no referencing Identifier present for a local
+
+    (cpgNodeCache.cachedIdentifier
+      .where(_.typeFullName(typeDeclVal))
+      .l ::: cpgNodeCache.cachedParameter.where(_.typeFullName(typeDeclVal)).l)
+      .foreach(impactedObject => {
+        if (impactedObject.tag.nameExact(Constants.id).l.isEmpty) {
+          storeForTag(builder, impactedObject, ruleCache)(
+            InternalTag.OBJECT_OF_SENSITIVE_CLASS_BY_MEMBER_NAME.toString,
+            taggerPart.ruleInfo.id
+          )
+          storeForTag(builder, impactedObject, ruleCache)(
+            Constants.id,
+            Constants.privadoDerived + Constants.underScore + cpgNodeCache.RANDOM_ID_OBJECT_OF_TYPE_DECL_HAVING_MEMBER_NAME
+          )
+          storeForTag(builder, impactedObject, ruleCache)(Constants.catLevelOne, CatLevelOne.DERIVED_SOURCES.name)
+        }
+        storeForTag(builder, impactedObject, ruleCache)(
+          Constants.privadoDerived + Constants.underScore + cpgNodeCache.RANDOM_ID_OBJECT_OF_TYPE_DECL_HAVING_MEMBER_NAME,
+          taggerPart.ruleInfo.id
+        )
+        // Tag for storing memberName in derived Objects -> user --> (email, password)
+        storeForTag(builder, impactedObject, ruleCache)(
+          taggerPart.ruleInfo.id + Constants.underScore + Constants.privadoDerived + Constants.underScore + cpgNodeCache.RANDOM_ID_OBJECT_OF_TYPE_DECL_HAVING_MEMBER_NAME,
+          typeDeclMemberName
+        )
+      })
+
+    // To Mark all field Access and getters
+    cpgNodeCache.tagAllFieldAccessAndGetters(builder, typeDeclVal, taggerPart.ruleInfo, typeDeclMemberName, true)
+  }
+}
+
+case class ExtendingDerivedSourceTaggerPart(typeDecl: TypeDecl, memberType: SecondLevelDerivedSourceTaggerPart)
+case class SecondLevelDerivedSourceTaggerPart(typeDecl: TypeDecl, ruleInfo: RuleInfo)
+case class DerivedSourceTaggerPart(typeDecl: TypeDecl, member: Member, ruleInfo: RuleInfo)
```

Comment:
Storing `ruleInfo` against each node can be a memory intensive operation, we ideally only need the `ruleId`

Reply:
No, it doesn't matter. Mapping RuleInfo vs ruleId will have a similar memory footprint, as it will only hold the reference to RuleInfo (we are not cloning the RuleInfo). The reference will occupy same size as that of Int.